browser/modules/SearchProvider.jsm

```diff
@@ -0,0 +1,260 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals Services, XPCOMUtils, Task, SearchSuggestionController, PrivateBrowsingUtils, FormHistory*/
+
+"use strict";
+
+this.EXPORTED_SYMBOLS = [
+  "SearchProvider",
+];
+
+const {
+  classes: Cc,
+  interfaces: Ci,
+  utils: Cu,
+} = Components;
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/Promise.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+XPCOMUtils.defineLazyModuleGetter(this, "FormHistory",
+  "resource://gre/modules/FormHistory.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "PrivateBrowsingUtils",
+  "resource://gre/modules/PrivateBrowsingUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "SearchSuggestionController",
+  "resource://gre/modules/SearchSuggestionController.jsm");
+
+const stringBundle = Services.strings.createBundle("chrome://global/locale/autocomplete.properties");
+const searchBundle = Services.strings.createBundle("chrome://browser/locale/search.properties");
+const MAX_LOCAL_SUGGESTIONS = 3;
+const MAX_SUGGESTIONS = 6;
+
+this.SearchProvider = {
+  // This is used to handle search suggestions.  It maps xul:browsers to objects
+  // { controller, previousFormHistoryResult }.  See getSuggestions().
+  _suggestionMap: new WeakMap(),
+
+  _searchSuggestionUIStrings: new Map(),
+  get searchSuggestionUIStrings() {
+    var result = {};
+    Array.from(this._searchSuggestionUIStrings.entries())
+      .forEach( ([key, value]) => result[key] = value);
+    return result;
+  },
+
+  get state() {
+    var t = Task.spawn(function* () {
+      let state = {
+        engines: [],
+        currentEngine: yield this.getCurrentEngine(),
+      };
+      let pref = Services.prefs.getCharPref("browser.search.hiddenOneOffs");
+      let hiddenList = pref ? pref.split(",") : [];
+      for (let engine of Services.search.getVisibleEngines()) {
+        if (hiddenList.indexOf(engine.name) !== -1) {
+          continue;
+        }
+        let uri = engine.getIconURLBySize(16, 16);
+        state.engines.push({
+          name: engine.name,
+          iconBuffer: yield this._arrayBufferFromDataURI(uri),
+        });
+      }
+      return state;
+    }.bind(this));
+
+    return t;
+  },
+
+  performSearch: function(browser, data) {
+    this._ensureDataHasProperties(data, [
```

Comment:
Gong to drop this. This is alien. Errors should occur if the properties are missing in place.